src/targets/gpu/fuse_ops.cpp

```diff
@@ -310,6 +331,116 @@ void move_standard_front(std::vector<instruction_ref>& args)
         std::swap(*it, args.front());
 }
 
+struct find_gelu
+{
+
+    static auto erf_fn()
+    {
+        return match::name("gpu::erf")(
+            match::used_once(),
+            match::arg(0)(match::used_once(),
+                          match::name("gpu::mul")(match::either_arg(0, 1)(
+                              match::none_of(match::has_value(M_SQRT1_2)).bind("x"),
+                              match::has_value(M_SQRT1_2)))));
+    }
+
+    auto matcher() const
+    {
+        return match::name("gpu::mul")(match::either_arg(0, 1)(
+            match::name("gpu::mul")(match::any_arg(0, 1)(match::args(match::has_value(0.5f)))),
+            match::name("gpu::add")(
+                match::used_once(),
+                match::either_arg(0, 1)(erf_fn(), match::args(match::has_value(1.0f))))));
+    }
+
+    void apply(program& p, match::matcher_result r) const
+    {
+        auto ins   = r.result;
+        auto x_ins = r.instructions["x"];
+        auto args  = ins->inputs();
+
+        p.replace_instruction(ins, hip_gelu{}, x_ins, args.back());
+    }
+};
+
+struct find_add_gelu
+{
+    auto matcher() const
+    {
+        return match::name("gpu::gelu")(match::arg(0)(match::name("gpu::add").bind("add")));
+    }
+
+    void apply(program& p, match::matcher_result r) const
+    {
+        auto add_ins = r.instructions["add"];
+        auto ins     = r.result;
+        auto args    = add_ins->inputs();
+        move_standard_front(args);
+        move_broadcasted_back(args);
+
+        args.back() = ins->inputs().back();
+        p.replace_instruction(ins, hip_add_gelu{}, args);
+    }
+};
+
+struct find_gelu_new
+{
+
+    static auto pow_fn()
+    {
+        return match::name("gpu::pow")(match::arg(1)(match::args(match::has_value(3.0f))));
+    }
+
+    static auto tanh_fn()
+    {
+        return match::name(
+            "gpu::tanh")(match::arg(0)(match::name("gpu::mul")(match::either_arg(0, 1)(
+            match::args(match::has_value(sqrt(M_2_PI))),
+            match::name("gpu::add")(match::any_arg(0, 1)(match::name("gpu::mul")(
+                match::either_arg(0, 1)(match::args(match::has_value(0.044715f)), pow_fn()))))))));
+    }
+
+    auto matcher() const
+    {
+        return match::name("gpu::mul")(match::either_arg(0, 1)(
+            match::any().bind("x"),
+            match::name("gpu::add")(match::any_arg(0, 1)(match::name("gpu::mul")(
+                match::either_arg(0, 1)(match::args(match::has_value(0.5f)), tanh_fn()))))));
```

Comment:
same for the tanh_fn(), do we need a used_once for it? Also for the gpu::mul here/

Reply:
changed